# Boosting is for Sequential Ensemple Learning

## Gradient Boosting is part of XGBoost

#__Boosting__
Next, let us understand the ensemble technique of boosting.

## Step 1: Import Required Libraries and Load the Dataset

- Import pandas, NumPy, sklearn.metrics, sklearn.model_selection, and sklearn.ensemble libraries



In [1]:
import pandas as pd
import numpy as np
from sklearn.metrics import classification_report
from sklearn.model_selection import KFold
from sklearn.datasets import load_breast_cancer
from sklearn.ensemble import GradientBoostingClassifier

- Load the breast cancer dataset and create a DataFrame df
- Assign the feature names of the dataset to columns and assign the target column to y
- Using the head() method, we can check the first 5 rows of the dataset. 

Let us load the data.

In [2]:
df = pd.DataFrame(load_breast_cancer()['data'],
columns=load_breast_cancer()['feature_names'])
df['y'] = load_breast_cancer()['target']
df.head(5)

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  y  
0          0.4601                  0.11890  0  
1          0.2750                  0.08902  0  
2          0.3613                  0.08758  0  
3          0.6638                  0.17300  0  
4          0.2364                  0.07678  0  

[5 rows x 31 columns]

__Observation:__
- This is the head of the dataset.

Let's check the data types and find information about the dataset.

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

__Observation:__
- There are no missing or null values.

## Step 2: Perform K-Fold Cross-Validation and Fit the Model
- Define X and y
- Assign the DataFrame df, except the target column, to X
- Assign the y column to variable y
- Create a K-fold
 - Iterate over the K-fold splits
- Split the data into training and validation sets

In [9]:
X = df.drop('y',axis=1)
y = df.y

#Get random splits
kf = KFold(n_splits=5,random_state=42,shuffle=True)

for train_index,val_index in kf.split(X):
    X_train,X_val = X.iloc[train_index],X.iloc[val_index],
    y_train,y_val = y.iloc[train_index],y.iloc[val_index],

- Create a GradientBoostingClassifier object with a learning rate of 0.1


In [10]:
# variation of gradient descent optimization (or stockastic gradient descent)
gradient_booster = GradientBoostingClassifier(learning_rate=0.1)

gradient_booster.get_params()

{'ccp_alpha': 0.0,
 'criterion': 'friedman_mse',
 'init': None,
 'learning_rate': 0.1,
 'loss': 'log_loss',
 'max_depth': 3,
 'max_features': None,
 'max_leaf_nodes': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_iter_no_change': None,
 'random_state': None,
 'subsample': 1.0,
 'tol': 0.0001,
 'validation_fraction': 0.1,
 'verbose': 0,
 'warm_start': False}

__Observation:__
- The above output shows the parameters that can be treated. In this demo, we apply gradient boosting with the default value.

Next, let's fit the model on the training data. 

In [11]:
gradient_booster.fit(X_train,y_train)

y_predict = gradient_booster.predict(X_val)

print(classification_report(y_val,y_predict))

              precision    recall  f1-score   support

           0       0.98      0.93      0.96        46
           1       0.96      0.99      0.97        67

    accuracy                           0.96       113
   macro avg       0.97      0.96      0.96       113
weighted avg       0.96      0.96      0.96       113



__Observations:__

- The accuracy is 0.96.
- Precision and recall do not have much difference between them.